win_dns_record

johnboy2 opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes ansible/ansible#51923

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
win_dns_record

##### ADDITIONAL INFORMATION
n/a

Commit:
win_dns: Initial work

Commit:
win_dns: initial commit

Commit:
Renaming win_dns -> win_dns_record

angystardust commented:
Hi @johnboy2 and thanks for this very useful module!
I've done some easy tests with ansible 2.7.5 and a Win2k12R2:
- adding an A or a CNAME record works fine but there's no idempotency: every time I run the play, the module emits a `changed` status instead of an `ok` :disappointed: 
- removing an A or a CNAME record works fine and the idempotency is satisfied :+1:

Commit:
win_dns_record: Fix record leakage in output

Commit:
win_dns_record: Fix erroneous minimums enforcement

It is apparently completely legitimate to specify a TTL that is below
minimum; it will just get ignored in favor of the server's minimum.

Commit:
win_dns_record: Fix new-host changes reported incorrectly

Commit:
win_dns_record: Fix TTL changes reported incorrectly

Commit:
win_dns_record: Fix existing records not recognized

Commit:
win_dns_record: Remove obsolete object

johnboy2 commented:
@angystardust - I believe I have that idempotency issue fixed.

it-praktyk commented:
The pull request don't contain integration tests :-(

angystardust commented:
@johnboy2 great! I'll take another "test tour" tomorrow ;)

angystardust commented:
So, I've done some tests again. Now idempotency is enforced for an A record but still gives a "changed" status for exiting CNAME records

Daniel-Sanchez-Fabregas committed:
Refactorize check mode

Daniel-Sanchez-Fabregas committed:
Add computer_name parameter

Daniel-Sanchez-Fabregas committed:
Refactorize diff and changed to read DNS end state

Daniel-Sanchez-Fabregas committed:
Fix pslint tests PSUseDeclaredVarsMoreThanAssignments and PSAvoidUsingCmdletAliases

Daniel-Sanchez-Fabregas committed:
Minor fix, misnamed variable.

Daniel-Sanchez-Fabregas committed:
win_dns_record: Fix "changed" state in check mode

Commit:
win_dns_record: cleanups

Commit:
win_dns_record: fix TTL update not changed regression

Commit:
Add initial integration tests

Commit:
win_dns_record: integration tests

Commit:
win_dns_record: Reverted 9cf5f2d8e6507cf477ab9e7ca166b1857169d6b5

The approach from that commit breaks check mode.

Commit:
win_dns_record: de-scope some records

These are either esoteric (meaning limited realworld testing)
or require additional thought to do properly (eg MX, which
has its "priority" level).

johnboy2 commented:
@Daniel-Sanchez-Fabregas - FYI I had to revert 9cf5f2d, as it breaks check-mode by omitting most of the (simulated) changes. Maybe that approach can be brought back in if we can find a better way to simulate the records.

dagwieers commented on lib/ansible/modules/windows/win_dns_record.ps1:
```diff
@@ -0,0 +1,148 @@
+# Copyright: (c) 2019, Hitachi ID Systems, Inc.
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+#Requires -Module Ansible.ModuleUtils.Legacy
```
This module is still using the legacy Powershell module interface.
We do have a new Powershell module interface which has many improvements.
Would you care to rewrite this module using the new interface?
Please look at some other Windows modules for examples.

johnboy2 replied:
This was written by following the [developer guide](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general_windows.html#windows-new-module-development), which still recommends the legacy module. My bias is to always studiously avoid stepping outside the documented box -- meaning I would prefer waiting for the developer guide to show some new path before jumping in.

varunchopra replied:
You're correct to go by the docs however that page has been updated for Ansible 2.8 and you can view it [here](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_general_windows.html#windows-new-module-development).

The new module interface was introduced in the latest developmental release and would be great if you could use this instead. You wouldn't have to change a lot of code to use it.

Daniel-Sanchez-Fabregas replied:
Tried to convert it in https://github.com/johnboy2/ansible/pull/2/commits/6a7d92a0791e5d58d11cd410d3a4c45d549820b7 . Tests are OK after conversion, but don know how to:
* ~~use `prepared` in `diff`~~
* use `required_if` in this context

Daniel-Sanchez-Fabregas replied:
I have this solved in https://github.com/johnboy2/ansible/pull/2 
Seems to me that `required_if` don't apply in this case.
Tests are OK

dagwieers commented on lib/ansible/modules/windows/win_dns_record.py:
```diff
@@ -0,0 +1,85 @@
+      - Specifies a DNS server. You can specify an IP address or any value that
+        resolves to an IP address, such as a fully qualified domain name
+        (FQDN), host name, or NETBIOS name.
+    required: false
```
If the parameter is not required, there is no need to add this.
```suggestion
```

dagwieers commented on lib/ansible/modules/windows/win_dns_record.py:
```diff
@@ -0,0 +1,85 @@
+version_added: "2.8"
+short_description: Manage Windows Server DNS records
+description:
+- Manages DNS records within an existing Windows Server DNS zone
```
All descriptions, except the short_description, must end with a trailing dot.
```suggestion
- Manages DNS records within an existing Windows Server DNS zone.
```

dagwieers commented on lib/ansible/modules/windows/win_dns_record.py:
```diff
@@ -0,0 +1,85 @@
+    type: str
+  ttl:
+    description:
+    - The "time to live" of the record, in seconds. Ignored when
```
Use multiple lines. The first line explains the parameter briefly.
Subsequent lines clarify when to use it, how to use it, what values are expected etc.
```suggestion
    - The "time to live" of the record, in seconds.
    - Ignored when
```

dagwieers commented on lib/ansible/modules/windows/win_dns_record.py:
```diff
@@ -0,0 +1,85 @@
+    type: str
+  zone:
+    description:
+    - The name of the zone to manage (eg C(example.com)). The zone must already
```
```suggestion
    - The name of the zone to manage (eg C(example.com)).
    - The zone must already
```

dagwieers commented on lib/ansible/modules/windows/win_dns_record.py:
```diff
@@ -0,0 +1,85 @@
+    type: str
+  computer_name:
+    description:
+      - Specifies a DNS server. You can specify an IP address or any value that
```
```suggestion
      - Specifies a DNS server.
      - You can specify an IP address or any value that
```

dagwieers reviewed (changes requested):
LGTM, although we may want to improve it further.

Daniel-Sanchez-Fabregas commented on test/integration/targets/win_dns_record/tasks/main.yml:
```diff
@@ -0,0 +1,19 @@
+  vars:
+    fail_on_missing: false
+
+- block:
```
Trying to replicate the tests locally breaks for me as DNS feature is missing. This solve it to me:
~~~ yaml
   - name: check if feature is availble
     win_shell: if (Get-Command -Name Add-WindowsFeature -ErrorAction SilentlyContinue) { $true } else { $false }
     changed_when: False
     register: module_available

   - name: install DNS features
     when: module_available.stdout | trim | bool
     block:

       - name: ensure DNS services are installed
         win_feature:
           name: DNS
           state: present
         register: dns_install

   - name: reboot server if needed
     win_reboot:
     when: dns_install.reboot_required
~~~

Daniel-Sanchez-Fabregas replied:
See https://github.com/johnboy2/ansible/pull/2

Daniel-Sanchez-Fabregas committed:
win_dns_records tests: Ensure DNS services are installed

Daniel-Sanchez-Fabregas commented on test/integration/targets/win_dns_record/tasks/tests-PTR.yml:
```diff
@@ -0,0 +1,186 @@
+  check_mode: yes
+
+- name: 'TYPE=PTR - creation get results (check mode)'
+  win_command: powershell.exe "If (Get-DnsServerResourceRecord -Zone '{{ win_dns_record_revzone }}' -Name '7' -RRType PTR -Node -ErrorAction:Ignore) { 'exists' } else { 'absent' }"
```
In order to make the tests work for meI have to substitute in this file all instances of `-Zone` by `-ZoneName` 
See https://github.com/johnboy2/ansible/pull/2

dagwieers committed:
Update lib/ansible/modules/windows/win_dns_record.py

Co-Authored-By: johnboy2

dagwieers committed:
Update lib/ansible/modules/windows/win_dns_record.py

Co-Authored-By: johnboy2

Commit:
Aggregated suggestions from dagwieers

Commit:
Fix bad powershell test argument

Daniel-Sanchez-Fabregas committed:
win_dns_record partially converted to new Powershell module interface

win_dns_record converted to new Powershell module interface, except diff 
and required_if

Daniel-Sanchez-Fabregas committed:
win_dns_record: convert diff support

Daniel-Sanchez-Fabregas committed:
win_dns_record: convert diff support to after-before style

Daniel-Sanchez-Fabregas committed:
Don't test for Add-WindowsFeature

Daniel-Sanchez-Fabregas committed:
win_dns_record: Fix diff

When check mode is used diff changes must be simulated.

Commit:
Style consistency/clean-ups

Commit:
Fix integration test typos

Commit:
Improve readability of diff output

The original intention of the diff output was to resemble zone file records
(except that the zone-name is added onto each record). In that light, the
missing *record class* information (always "IN" in our case) was an oversight.

This just makes the diff output more "instantly readable" for DNS gurus.

Commit:
win_dns_record: Add diff tests

johnboy2 commented:
@dagwieers - Your review points were applied, but remain outstanding in the UI. Can you dismiss and/or re-review?

Commit:
Fix ansible-test sanity check fails

angystardust commented:
@dagwieers @johnboy2 no chance to see this one in 2.8 GA?

Daniel-Sanchez-Fabregas commented:
Maybe 951164ec4915e318f30e63b02c7cff83a5345fd7 was needed https://app.shippable.com/github/ansible/ansible/runs/112013/17/tests
@johnboy2 could you try to pass the tests adding it?

dagwieers commented on lib/ansible/modules/windows/win_dns_record.ps1:
```diff
@@ -0,0 +1,149 @@
+        type = @{ type = "str"; choices = "A","AAAA","CNAME","PTR"; required = $true }
+        value = @{ type = "list"; elements = "str"; default = @() ; aliases=@( 'values' )}
+        zone = @{ type = "str"; required = $true }
+        computer_name = @{ type = "str"; required = $false }
```
If it is not required, it is not needed.
```suggestion
        computer_name = @{ type = "str" }
```

dagwieers commented on lib/ansible/modules/windows/win_dns_record.ps1:
```diff
@@ -0,0 +1,149 @@
+
+
+$extra_args = @{}
+if ($dns_computer_name -ne $null) {
```
Thanks to PowerShell's design...
```suggestion
if ($null -ne $dns_computer_name) {
```

dagwieers commented on lib/ansible/modules/windows/win_dns_record.ps1:
```diff
@@ -0,0 +1,149 @@
+
+
+$records = Get-DnsServerResourceRecord -ZoneName $zone -Name $name -RRType $type -Node -ErrorAction:Ignore @extra_args | Sort-Object
+if ($records -ne $null) {
```
```suggestion
if ($null -ne $records) {
```

dagwieers commented on lib/ansible/modules/windows/win_dns_record.ps1:
```diff
@@ -0,0 +1,149 @@
+
+if ($state -eq 'present') {
+    if ($values.Count -eq 0) {
+        $module.FailJson("values must be non-empty when state='present'")
```
Capitalize error messages. (You have some capitalized in this script, so keep it consistent)

johnboy2 replied:
I wasn't sure how to deal with that one given that "values" is an input variable, but "Values" is not.

johnboy2 replied:
Actually I'm just going to reword it so the first word isn't the attribute name.

johnboy2 replied:
I've reworked the affected sentences instead.

dagwieers commented on lib/ansible/modules/windows/win_dns_record.ps1:
```diff
@@ -0,0 +1,149 @@
+    }
+} else {
+    if ($values.Count -ne 0) {
+        $module.FailJson("values must be undefined or empty when state='absent'")
```
Capitalize.

johnboy2 replied:
Ditto my last.

dagwieers commented on lib/ansible/modules/windows/win_dns_record.py:
```diff
@@ -0,0 +1,85 @@
+version_added: "2.8"
+short_description: Manage Windows Server DNS records
+description:
+- Manages DNS records within an existing Windows Server DNS zone.
```
```suggestion
- Manage DNS records within an existing Windows Server DNS zone.
```

dagwieers commented on lib/ansible/modules/windows/win_dns_record.ps1:
```diff
@@ -0,0 +1,149 @@
+$ttl = New-TimeSpan -Seconds $ttl
+
+
+if (($type -eq 'CNAME' -or $type -eq 'PTR') -and $values -ne $null -and $values.Count -gt 0 -and $zone[-1] -ne '.') {
```
```suggestion
if (($type -eq 'CNAME' -or $type -eq 'PTR') -and $null -ne $values -and $values.Count -gt 0 -and $zone[-1] -ne '.') {
```

dagwieers commented on lib/ansible/modules/windows/win_dns_record.ps1:
```diff
@@ -0,0 +1,149 @@
+}
+
+
+if ($values -ne $null -and $values.Count -gt 0) {
```
```suggestion
if ($null -ne $values -and $values.Count -gt 0) {
```

dagwieers reviewed (changes requested):
LGTM, a few changes.

I would also like to see full support for all record types, but if we can start with only a subset, that's fine for me too.

dagwieers commented:
@angystardust I hope so. It would help if more people review and test the code, and add a shipit.

dagwieers commented:
The tests seem to fail on Windows 2008 and 2008R2, so probably should be excluded from running on those.

angystardust commented:
@dagwieers thanks for your review. I'm already using the module in the current status on a 2.7 setup and..."it works for me" :copyright: 
Happy to test further if a newer revision is pushed ;)

dagwieers committed:
Apply suggestions from code review

Added suggestions from dagwieers

Co-Authored-By: johnboy2 <john@jnelson.ca>

Commit:
win_dns_record: Skip 2008 and friends

Commit:
Reword error messages so they start capitalized.

Commit:
Fix sanity error

johnboy2 commented:
> Maybe [951164e](https://github.com/ansible/ansible/commit/951164ec4915e318f30e63b02c7cff83a5345fd7) was needed https://app.shippable.com/github/ansible/ansible/runs/112013/17/tests
> @johnboy2 could you try to pass the tests adding it?

My concerns with that idea are that it leaves the feature uninstalled if it won't work, and that I shouldn't have to work around functionality already provided by (and presumably tested for) the `win_feature` module.

A few other things:
 * @dagwieers' suggestions applied. For the capitalization ones, the affected sentences have been reworked.
 * RE failing tests, looks like Win2008/etc don't support the essential cmdlets used by this module. So I've also added it to the skip list for those platforms (also @dagwieers' suggestion) plus added that fact to the module documentation.
 * RE supporting other record types, I agree completely but didn't have time to even think about other types, much less test them, consider corner cases, etc. For me at least, priority <span>#</span>1 was getting something workable out the door.

jborean93 commented on lib/ansible/modules/windows/win_dns_record.ps1:
```diff
@@ -0,0 +1,149 @@
+
+
+# TODO: add warning for forest minTTL override -- see https://docs.microsoft.com/en-us/windows/desktop/ad/configuration-of-ttl-limits
+if ($ttl -lt 1 -or $ttl -gt 31557600) {
```
This should be documented in the option description.

johnboy2 replied:
Agreed -- I'll add that range limit.

johnboy2 replied:
Done.

jborean93 commented on test/integration/targets/win_dns_record/tasks/main.yml:
```diff
@@ -0,0 +1,30 @@
+  vars:
+    fail_on_missing: false
+
+- block:
```
While you've added the skip in the aliases, you should still check and skip if necessary when running on older versions of Windows. Usually it's a simple

```
- win_shell: if (Get-Command -Name ... -ErrorAction SilentlyContinue) { $true } else { $false }
  register: run_tests

- name: run tests on supported hosts
  when: run_tests.stdout | trim | bool
  block:
```

johnboy2 replied:
Alright, I know when I'm beat (@Daniel-Sanchez-Fabregas mentioned something similar).

But keep in mind that the relevant cmdlets aren't available _even on supported platforms_ unless/until the DNS feature is installed.

So I've added the cmdlet tests, but also put OS-version tests in front of them to avoid installing the DNS feature unnecessarily. That has the effect of making the cmdlet test redundant, however.

jborean93 replied:
Ah sorry that makes sense, I would just do the OS check then, sorry for the confusion.

johnboy2 replied:
Done.

jborean93 reviewed (changes requested):
Looks great, don't know much about DNS but you've covered pretty much all of the basics.

Commit:
win_dns_record: Document ttl range

Commit:
win_dns_record: Additional supportability barriers in tests

Commit:
win_dns_record: Typo

Commit:
win_dns_record: Sanity fix

Commit:
win_dns_record: Use OS-test only for compat checking

Daniel-Sanchez-Fabregas commented on test/integration/targets/win_dns_record/tasks/tests.yml:
```diff
@@ -1,3 +1,13 @@
+
+- name: reboot server if needed
+  win_reboot:
+  when: dns_install.reboot_required
```
@johnboy2. Maybe better Asking for changes here? like:

https://github.com/ansible/ansible/blob/0e980596376be7b7ef843d20b52228687e94717d/test/integration/targets/win_dns_record/tasks/main.yml#L16-L18

Daniel-Sanchez-Fabregas replied:
Ooops, you just changed it. :innocent: Though this was other file.

johnboy2 replied:
:-)

johnboy2 commented:
@jborean93 @dagwieers - can you re-review?

jborean93 reviewed (approved):
Looks great, thanks for working on this one.
